src/Features/Core/Portable/Navigation/INavigableItem.cs

```diff
@@ -45,5 +48,39 @@ internal interface INavigableItem
         bool IsStale { get; }
 
         ImmutableArray<INavigableItem> ChildItems { get; }
+
+        public record NavigableDocument(NavigableProject Project, string Name, string? FilePath, IReadOnlyList<string> Folders, DocumentId Id)
+        {
+            public required bool IsSourceGeneratedDocument { get; init; }
+            public required Workspace Workspace { get; init; }
+
+            public static NavigableDocument FromDocument(Document document)
+                => new(NavigableProject.FromProject(document.Project), document.Name, document.FilePath, document.Folders, document.Id)
+                {
+                    IsSourceGeneratedDocument = document is SourceGeneratedDocument,
+                    Workspace = document.Project.Solution.Workspace,
+                };
+
+            internal ValueTask<Document> GetDocumentAsync(Solution solution, CancellationToken cancellationToken)
+                => solution.GetRequiredDocumentAsync(Id, includeSourceGenerated: IsSourceGeneratedDocument, cancellationToken);
```

Comment:
this doesn't fele safe.  Since we're not rooting the solution, then items may refer to docs that now no longer exist.  we should reflect that in this api and make sure downstream consumers are resilient to that happening.

Reply:
Wouldn't a failure here require the deletion of a document containing a Navigate To result _while the Navigate To UI was active_? I can try to make it not required though.

Reply:
➡️ There was only one call to `GetDocumentAsync` (directly or indirectly) which didn't pass in the same solution used to create the items in the first place. It was the call to `GetTextSynchronously`, which has now been updated to `TryGetTextSynchronously` and the caller updated to handle the case where the document was no longer available.

Reply:
The navto system (now called AIOSP) supports pinning the results, and going back to them and interacting with them *post* search.  So it would be possible to potentially run into a doc that no longer exists.  Def wonky for sure.  I'm ok if you punt on that.  just a note in the code would be good.

Reply:
i support this given that this passes in the original solution.  Can you *doc* this though that's mandatory for this to be safe. 

I think you should also rename this to GetRequiredDocumentAsync to make it clear it will die if not there, and that's  a bug in the caller for passing bad data.

If you do the above, i'm fine with this approach :)

Reply:
➡️ Documented and renamed